README.md
# Forschungsdatenfreigabe elektronisch Patientenakte

Wie in [gemF\_ePA\_FDZ\_Anbindung](https://fachportal.gematik.de)
überblicksartig dargstellt, können bei Einwilligung der versicherten Person auf
dem ePA-Frontend des Versicherten (ePA-FdV) medizinische Daten zunächst
anonymisiert werden und diese dann mit Hilfe einer Vertrauensstelle (VST,
verantwortet vom RKI) pseudonymisiert werden. Diese pseudonymisierten
medizinischen Daten werden dann vom ePA-FdV an das Forschungsdatenzentrum (FDZ)
gesendet.

## Beispiel-Implementierung Schnittstellen

Vertrauensstelle: [vst.py](vst.py),
[Erläuterungen zur Schnittstelle](README-VST.md),
[VST-OpenAPI-Spezifikation (YAML)](openapi/openapi-vst.yaml).

Forschungsdatenzentrum: [fdz.py](vst.py),
[Erläuterungen zur Schnittstelle](README-FDZ.md),
[FDZ-OpenAPI-Spezifikation (YAML)](openapi/openapi-fdz.yaml). 

ePA Frontend des Versicherten: [epa-fdv.py](epa-fdv.py)

ePA Aktensystem: [as.py](as.py) (Notwendig für die Erzeugung der Auftragsnummer (AN), des Lieferpseudonyms (LP) und des Proxy-Tokens)

### Docker-Image

Eine lokale Instanz kann man auch leicht mit
dem [Dockerimage](https://hub.docker.com/r/gematik1/epa-forschungsdatenfreigabe-poc) starten

    docker pull gematik1/epa-forschungsdatenfreigabe-poc:1.1.0 
    docker run -it gematik1/epa-forschungsdatenfreigabe-poc:1.1.0 /bin/bash

Dann hat man eine interaktive Shell in der man eingibt:

    ./start.sh
    ./epa-fdv

Toll.

Wenn man tiefer einsteigen möchte, mit

    tmux at

in die tmux-session gehen und mit Ctrl+A plus 1 bis 4 zwischen den Fenstern
hin und her schalten. D. h., `Ctrl+A 1` geht ins das erste Fenstern etc.


### Komponenten/Services per "Hand" starten

Wer nicht das Docker-Image verwenden möchte, kann die REST-Services auch per "Hand" starten:

    ./as.py
    ./vst.py
    ./fdz.py


### Beispieldurchlauf
    $> ./epa-fdv.py
    Beim AS (http://127.0.0.1:20001/datenfreigabe/token) (1) ein signiertes AN+LP-Tupel und  (2) eine signierte AN beziehen.

    Ergebnis: (1) eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCIsIng1YyI6WyJNSUlDQWpDQ0Fha0
    NGRmUyTVRZVXJYMjhVN2FYSm9XemIrWEYzcTlyTUFvR0NDcUdTTTQ5QkFNQ01IQXhDekFKQmdOVkJBWV
    RBa1JGTVE4d0RRWURWUVFJREFaQ1pYSnNhVzR4RHpBTkJnTlZCQWNNQmtKbGNteHBiakVRTUE0R0ExVU
    VDZ3dIWjJWdFlYUnBhekVRTUE0R0ExVUVDd3dIWjJWdFlYUnBhekViTUJrR0ExVUVBd3dTUzI5dGNHOX
    VaVzUwWlc0dFVFdEpMVU5CTUI0WERUSXlNRE13TVRFek5USXpPRm9YRFRJek1ETXdNVEV6TlRJek9Gb3
    dnWll4Q3pBSkJnTlZCQVlUQWtSRk1ROHdEUVlEVlFRSURBWkNaWEpzYVc0eER6QU5CZ05WQkFjTUJrSm
    xjbXhwYmpFWU1CWUdBMVVFQ2d3UFpWQkJJRUZyZEdWdWMzbHpkR1Z0TVNFd0h3WURWUVFMREJoTGIyMX
    diMjVsYm5SbElFRjFkRzl5YVhOcFpYSjFibWN4S0RBbUJnTlZCQU1NSDJWUVFTMUJVeUJMYjIxd2IyNW
    xiblJsSUVGMWRHOXlhWE5wWlhKMWJtY3dXakFVQmdjcWhrak9QUUlCQmdrckpBTURBZ2dCQVFjRFFnQU
    VraytOUDY4QWVVMXR6cis1UEFVZWtwWjNyTGtqU2h2ODhVSkpEQkFTSDh4emlYTU84TW1aU0p5SHZXRn
    g0NlB4Rm80SU5CZmErcmh2dXBUTkhmbm9WakFLQmdncWhrak9QUVFEQWdOSEFEQkVBaUI5ZXpNMEtBck
    IwWU9XVi9YeGV6eDhYNTVtdi9ZN1Rld2dxRUJSS0psV2xnSWdJckdZa1QvdStZWjdieStVOE5lVHAzN2
    8xWkZRMkJ2NStLL3paZGlmQUFjPSJdfQ.eyJ3biI6IkZYRkYwSW5rY3o4Q09CY3hiV1l5Y1Y0MDlERnZ
    pTUMxeldpUlgrYW5ObXM9IiwiZHAiOiJBVjljQmFMSVlHRmJuakdPK3kyMENUenBFeWNuSFpRaDQxL3l
    GRlVyNGhERkVjMjZiOGlFb3Ewd0g4UWFtbkdQRFZscm0vQUFoNFp3L1VMY2pWa3JobGoyaHhONkY2WGM
    4MzdCbDhVaHNUMXZyVTE4aHRBZXkzYWxZZVdLMld4QXpVVERubTBPaFE9PSIsImlhdCI6MTY2NjA4OTM
    2MC41MzcxOTYsImV4cCI6MTY2NjE3NTc2MC41MzcxOTYsImlzcyI6Imh0dHA6Ly9hdXRob3JpemF0aW9
    uLmFrdGVuc3lzdGVtLnRpIiwidnN0X2NlcnRfaGFzaCI6IkJtcjM1TkQzdzVrZW9LMVF0VVNLNmNrRC9
    1M3RBYXJ1NkhuaEYxTThWNnM9In0.WAKzoVGaIOGvZXde9J5SsBMC0VFPirLTcobAfFlYEhVXyWe2Uld
    -dA5RSbwZBrhYaRVCrWYW1UXVmRqW0rnklA

    Ergebnis: (2) eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCIsIng1YyI6WyJNSUlDQWpDQ0Fha0
    NGRmUyTVRZVXJYMjhVN2FYSm9XemIrWEYzcTlyTUFvR0NDcUdTTTQ5QkFNQ01IQXhDekFKQmdOVkJBWV
    RBa1JGTVE4d0RRWURWUVFJREFaQ1pYSnNhVzR4RHpBTkJnTlZCQWNNQmtKbGNteHBiakVRTUE0R0ExVU
    VDZ3dIWjJWdFlYUnBhekVRTUE0R0ExVUVDd3dIWjJWdFlYUnBhekViTUJrR0ExVUVBd3dTUzI5dGNHOX
    VaVzUwWlc0dFVFdEpMVU5CTUI0WERUSXlNRE13TVRFek5USXpPRm9YRFRJek1ETXdNVEV6TlRJek9Gb3
    dnWll4Q3pBSkJnTlZCQVlUQWtSRk1ROHdEUVlEVlFRSURBWkNaWEpzYVc0eER6QU5CZ05WQkFjTUJrSm
    xjbXhwYmpFWU1CWUdBMVVFQ2d3UFpWQkJJRUZyZEdWdWMzbHpkR1Z0TVNFd0h3WURWUVFMREJoTGIyMX
    diMjVsYm5SbElFRjFkRzl5YVhOcFpYSjFibWN4S0RBbUJnTlZCQU1NSDJWUVFTMUJVeUJMYjIxd2IyNW
    xiblJsSUVGMWRHOXlhWE5wWlhKMWJtY3dXakFVQmdjcWhrak9QUUlCQmdrckpBTURBZ2dCQVFjRFFnQU
    VraytOUDY4QWVVMXR6cis1UEFVZWtwWjNyTGtqU2h2ODhVSkpEQkFTSDh4emlYTU84TW1aU0p5SHZXRn
    g0NlB4Rm80SU5CZmErcmh2dXBUTkhmbm9WakFLQmdncWhrak9QUVFEQWdOSEFEQkVBaUI5ZXpNMEtBck
    IwWU9XVi9YeGV6eDhYNTVtdi9ZN1Rld2dxRUJSS0psV2xnSWdJckdZa1QvdStZWjdieStVOE5lVHAzN2
    8xWkZRMkJ2NStLL3paZGlmQUFjPSJdfQ.eyJ3biI6IkZYRkYwSW5rY3o4Q09CY3hiV1l5Y1Y0MDlERnZ
    pTUMxeldpUlgrYW5ObXM9IiwiaWF0IjoxNjY2MDg5MzYwLjUzNzE5NiwiZXhwIjoxNjY2MTc1NzYwLjU
    zNzE5NiwiaXNzIjoiaHR0cDovL2F1dGhvcml6YXRpb24uYWt0ZW5zeXN0ZW0udGkifQ.i52vCa5m_T5y
    oHWqFdR8T-U5T5MUo9HweHr50Ah8Lcx-iy9zsw-128N78x6b7qUQApTNg3CZNiOtpqbWDb7VOQ 

    Im Body von (1) befindet sich:
    {
        "dp": "AV9cBaLIYGFbnjGO+y20CTzpEycnHZQh41/yFFUr4hDFEc26b8iEoq0wH8QamnGPDVlrm/AAh4Zw/ULcjVkrhlj2hxN6F6Xc837Bl8UhsT1vrU18htAey3alYeWK2WxAzUTDnm0OhQ==",
        "exp": 1666175760.537196,
        "iat": 1666089360.537196,
        "iss": "http://authorization.aktensystem.ti",
        "vst_cert_hash": "Bmr35ND3w5keoK1QtUSK6ckD/u3tAaru6HnhF1M8V6s=",
        "wn": "FXFF0Inkcz8COBcxbWYycV409DFviMC1zWiRX+anNms="
    }

    Das signierte Tupel (s. o.) sende ich jetzt an die Vertrauensstelle.
    Response-Code: 201
    Ergebnis:
    {'status': 'ok'}

    Jetzt verschlüssele ich mein MIO für das FDZ und sende es an das FDZ inkl. AN
    Hex-Dump des Chiffrats was ich gleich auf http://127.0.0.1:20003/v1/epa/mio/55682e888191984f833627c09893a1e4807bcd18ad83ba60c2ecc18312b9e257 POST-te:
    01628758dc762332804a54d91382e9115bf4549d14b7fea750b4cd318ee29555a800040016a4
    22b03d85d7d53859b54eeed07b005e0b732f2a9d61c2973590c2768c540bb9c5318df66095485dcd
    eb1807bc934b45fd063e155d44ee049a27f9b7a0c414b7b6c598e4eb1789829a0b4d068df030d0fb
    547c33d29453313974585cf3c5ac87eafc18337a07bfd5ebd7fe685868378bb1e9b155d42e2acc93
    1261d1e55dcfe0c46c8cb8550bd7e6641d00f6094683d3a49c21a0be40828e271bbd46f020ec9e

    Response-Code: 201
    Ergebnis:
    {'status': 'ok'}

    Jetzt widerrufe ich die Datenfreigabe wieder für das eben freigegebene MIO
    Response-Code: 202
    Ergebnis:
    {'status': 'ok'}

    Jetzt widerrufe ich als Test viele MIOs auf einmal
    Response-Code: 202
    Ergebnis:
    {'status': 'ok'}

# License

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "
AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
language governing permissions and limitations under the License.


ReleaseNotes.md
# Release 1.1.0
- corrected ECDSA signature (jws). Now conforms to rfc-7515

# Release 1.0.0
Version 1.0.0



epa-fdv.py
#! /usr/bin/env python3
# -*- coding: UTF-8 -*-

import base64, json, hashlib, requests, secrets, sys, time
from datetime import datetime, timedelta

from binascii import hexlify, unhexlify

from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def jb64d(data):
    """
    Adds back in the required padding before decoding.
    """
    padding_len = 4 - (len(data) % 4)
    data = data + ("=" * padding_len)
    return base64.urlsafe_b64decode(data)

if __name__ == '__main__':

    AS_URL_1 = "http://127.0.0.1:20001/datenfreigabe/token"

    # Ich nehme im Beispiel an, dass ich als ePA-FdV schon am AS vollständig
    # eingeloggt bin.

    print(f"Beim AS ({AS_URL_1}) (1) ein signiertes AN+LP-Tupel und",
          " (2) eine signierte AN beziehen.")

    as_server_online = False
    while not as_server_online:
        try:
            response = requests.get(
                            AS_URL_1,
                            headers={"SOAP_Access_Token" : "...AC-Token..."},
                            timeout=5
                        )
        # https://stackoverflow.com/questions/16511337/correct-way-to-try-except-using-python-requests-module
        except requests.exceptions.RequestException:
            print("AS HTTP Server ist noch nicht vollständig gestartet ... 5 Sekunden warten")
            time.sleep(5)
        else:
            as_server_online = True

    assert response.status_code == requests.codes.ok
    [signed_wn_dp_tupel, signed_wn] = response.json(); 
    print("\nErgebnis: (1)", signed_wn_dp_tupel)
    print("\nErgebnis: (2)", signed_wn, "\n")

    # Nur mal als Info den Body des signierten LP+AN-Tupel anzeigen
    body_data = signed_wn_dp_tupel.split(".")[1]
    body = json.loads(jb64d(body_data))
    print("Im Body von (1) befindet sich:\n", json.dumps(body, sort_keys=True, indent=4), sep='')

    print("\nDas signierte Tupel (s. o.) sende ich jetzt an die Vertrauensstelle.")
    VST_URL = "http://127.0.0.1:20002/v1/epa/signed_wn_dp"
    response = requests.post(
                    VST_URL,
                    headers={'Content-Type': 'application/jwt'},
                    data=signed_wn_dp_tupel,
                    timeout=5
                )
    print("Response-Code:", response.status_code)
    assert response.status_code == 201

    j = response.json(); 
    print("Ergebnis:\n", j, sep='')

    print("\nJetzt verschlüssele ich mein MIO für das FDZ und sende es an das FDZ inkl. AN")
    # für ein Datenobjekt erzeugt das ePA-FdV zufällig einen 256-Bit data oject id
    p_document_id = secrets.token_hex(32)

    FDZ_URL = f"http://127.0.0.1:20003/v1/epa/mio/{p_document_id}"

    with open("data/mio_1.txt", "rb") as f:
        plaintext = f.read()
        assert plaintext

    with open("pki/fdz.der", "rb") as f:
        fdz_cert_der = f.read()
        fdz_cert_hash = base64.b64encode(hashlib.sha256(fdz_cert_der).digest()).decode()
        fdz_cert = x509.load_der_x509_certificate(fdz_cert_der, default_backend())
        # das geladene Zertifikat muss jetzt in einer echten Implementierung 
        # erst noch geprüft werden bevor der Schlüssel aus dem Zertifikat 
        # wirklich benutzt wird ... nicht Teil des PoCs.
        fdz_public_key = fdz_cert.public_key()

    private_key = ec.generate_private_key(ec.BrainpoolP256R1(), default_backend())
    pn = private_key.public_key().public_numbers()
    shared_secret = private_key.exchange(ec.ECDH(), fdz_public_key)
    hkdf = HKDF(algorithm=hashes.SHA256(), length=32, salt=None, info=b'', backend=default_backend())
    aes_key = hkdf.derive(shared_secret)
    iv = secrets.token_bytes(12)
    ciphertext_aes = AESGCM(aes_key).encrypt(iv, plaintext, associated_data=None)
    x = unhexlify("{:x}".format(pn.x).zfill(64))
    y = unhexlify("{:x}".format(pn.y).zfill(64))

    ciphertext = b'\x01' + x + y + iv + ciphertext_aes

    print(f"Hex-Dump des Chiffrats was ich gleich auf {FDZ_URL} POST-te:")
    print(hexlify(ciphertext).decode())

    response = requests.post(
                    FDZ_URL,
                    headers={'Signed-WN'     : signed_wn,
                             'FDZ-Cert-Hash' : fdz_cert_hash,
                             'Content-Type'  : 'application/octet-stream'},
                    data=ciphertext,
                    timeout=5
                )
    print("Response-Code:", response.status_code)
    assert response.status_code == 201

    j = response.json(); 
    print("Ergebnis:\n", j, sep='')

    print("\nJetzt widerrufe ich die Datenfreigabe wieder für das eben freigegebene MIO")

    response = requests.delete(
                    FDZ_URL,
                    headers={'Signed-WN'     : signed_wn},
                    timeout=5
                )

    print("Response-Code:", response.status_code)
    assert response.status_code in [200, 202]

    j = response.json(); 
    print("Ergebnis:\n", j, sep='')

    print("\nJetzt widerrufe ich als Test viele MIOs auf einmal")

    FDZ_URL = f"http://127.0.0.1:20003/v1/epa/mios"

    response = requests.delete(
                    FDZ_URL,
                    headers={'Signed-WN'     : signed_wn},
                    json = [secrets.token_hex(32) for _ in range(0,10)],
                    timeout=5
                )
    print("Response-Code:", response.status_code)
    assert response.status_code in [200, 202]

    j = response.json(); 
    print("Ergebnis:\n", j, sep='')

